backend/moonstream/providers/bugout.py

```diff
@@ -0,0 +1,273 @@
+"""
+Event providers powered by Bugout journals.
+"""
+from datetime import datetime
+import json
+import logging
+from typing import Dict, List, Optional, Tuple
+
+from bugout.app import Bugout
+from bugout.data import BugoutResource, BugoutSearchResult
+from bugout.journal import SearchOrder
+from dateutil.parser import isoparse
+from dateutil.tz import UTC
+from sqlalchemy.orm import Session
+
+from .. import data
+from ..stream_queries import StreamQuery
+
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.WARN)
+
+
+class BugoutEventProviderError(Exception):
+    """
+    Catch-all error for BugoutEventProvider instances.
+    """
+
+
+class BugoutEventProvider:
+    """
+    Provides events (specified by a conjunction of tags) from a Bugout journal.
+    """
+
+    def __init__(
+        self,
+        event_type: str,
+        tags: Optional[List[str]] = None,
+        batch_size: int = 100,
+        timeout: float = 30.0,
+    ):
+        """
+        Args:
+        - event_type: Name of event this instance provides
+        - tags: Tags which define events that this provider works with
+        - batch_size: Number of events to read from journal at a time
+        - timeout: Request timeout for Bugout requests
+        """
+        self.event_type = event_type
```

Comment:
@Andrei-Dolgolev : I think this is a good idea - we can add the `token` and `journal_id` as arguments to `BugoutEventProvider`. This will make it easy for third parties to offer data through the Moonstream dashboard.

Reply:
Created an issue for this: https://github.com/bugout-dev/moonstream/issues/157